### ISSUE TITLE
tf.keras.metrics.MeanIoU outcome is not improving

### ISSUE THREAD

**Issue by mehran66** [tensorflow/tensorflow#55565]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.8.0
- Python version: 3.95

**Describe the current behavior**
I am using tf.keras.metrics.MeanIoU as a metric in a semantic segmentation problem. All of the other metrics (accuracy, precision, recall) are changing during training but MeanIoU does not change. I am using tf.keras.losses.CategoricalCrossentropy(from_logits=False) for my loss function.

This issue has been reported in the following two StackOverFlow but no solution provided yet.

https://stackoverflow.com/questions/71729853/tf-keras-metrics-meaniou-outcome-is-not-improving/71813168#71813168

https://stackoverflow.com/questions/70848143/mean-iou-in-tensorflow-not-updating-resulting-in-correct-value/71813135#71813135

Here is an example of metrics during training:

Epoch 1/30

Epoch 1: val_loss improved from inf to 2.39962, saving model to /content/drive/MyDrive/Colab/test/semantic_segmentation/models_trained/training-min-val_loss.hdf5
269/269 - 899s - loss: 0.2277 - accuracy: 0.9108 - IoU: 0.2500 - val_loss: 2.3996 - val_accuracy: 0.9286 - val_IoU: 0.8372 - lr: 0.0010 - 899s/epoch - 3s/step
Epoch 2/30

Epoch 2: val_loss did not improve from 2.39962
269/269 - 853s - loss: 0.1805 - accuracy: 0.9320 - IoU: 0.2500 -  val_loss: 4.5132 - val_accuracy: 0.9244 - val_IoU: 0.8303 -  lr: 0.0010 - 853s/epoch - 3s/step
Epoch 3/30

Epoch 3: val_loss did not improve from 2.39962
269/269 - 817s - loss: 0.1672 - accuracy: 0.9380 - IoU: 0.2500 -  val_loss: 3.2384 - val_accuracy: 0.9198 - val_IoU: 0.8092 - lr: 0.0010 - 817s/epoch - 3s/step
Epoch 4/30

Epoch 4: val_loss improved from 2.39962 to 0.36882, saving model to /content/drive/MyDrive/Colab/test/semantic_segmentation/models_trained/training-min-val_loss.hdf5
269/269 - 855s - loss: 0.1643 - accuracy: 0.9390 - IoU: 0.2500 - val_loss: 0.3688 - val_accuracy: 0.9337 - val_IoU: 0.2600 -  lr: 0.0010 - 855s/epoch - 3s/step

If I change my one hot encoded labels to a one band maks, use SparseCategoricalAccuracy and the following modified MeanIoU....it works all fine.

class SparseMeanIoU(tf.keras.metrics.MeanIoU):
  def __init__(self,
               y_true=None,
               y_pred=None,
               num_classes=None,
               name=None,
               dtype=None):
    super(SparseMeanIoU, self).__init__(num_classes = num_classes,name=name, dtype=dtype)

  def update_state(self, y_true, y_pred, sample_weight=None):
    y_pred = tf.math.argmax(y_pred, axis=-1)
    return super().update_state(y_true, y_pred, sample_weight)

**Describe the expected behavior**
MeanIoU should change during training.

- Do you want to contribute a PR? (yes/no): no

**Comment by mehran66** [comment]
I found the solution ... for tf.keras.losses.CategoricalCrossentropy, I should use tf.keras.metrics.OneHotMeanIoU

**Comment by sushreebarsa** [comment]
@mehran66 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here if this is still an issue? 
Please move this issue to closed status if it is resolved for you?Thanks!
